controllers/integration/integration_adapter.go

```diff
@@ -137,6 +137,136 @@ func (a *Adapter) getAllApplicationSnapshots() (*[]appstudioshared.ApplicationSn
 }
 
+func (a *Adapter) getAllEnvironments() (*[]appstudioshared.Environment, error) {
+
+	environmentList := &appstudioshared.EnvironmentList{}
+	opts := []client.ListOption{
+		client.InNamespace(a.application.Namespace),
+	}
+	err := a.client.List(a.context, environmentList, opts...)
+	return &environmentList.Items, err
+}
+
+func (a *Adapter) findAvailabeEnvironments() (*[]appstudioshared.Environment, error) {
+	allEnvironments, err := a.getAllEnvironments()
+	if err != nil {
+		return nil, err
+	}
+	availableEnvironments := []appstudioshared.Environment{}
+	for _, environment := range *allEnvironments {
+		if environment.Spec.ParentEnvironment == "" {
+			availableEnvironments = append(availableEnvironments, environment)
+		} else {
+			for _, tag := range environment.Spec.Tags {
+				if tag == "ephemeral" {
+					availableEnvironments = append(availableEnvironments, environment)
+				}
+			}
+		}
+	}
+	return &availableEnvironments, nil
+}
+
+func (a *Adapter) findExistingApplicationSnapshotEnvironmentBinding(environmentName string) (*appstudioshared.ApplicationSnapshotEnvironmentBinding, error) {
+	applicationSnapshotEnvironmentBindingList := &appstudioshared.ApplicationSnapshotEnvironmentBindingList{}
+	opts := []client.ListOption{
+		client.InNamespace(a.application.Namespace),
+		client.MatchingFields{"spec.application": a.application.Name},
+		client.MatchingFields{"spec.environment": environmentName},
+	}
+
+	err := a.client.List(a.context, applicationSnapshotEnvironmentBindingList, opts...)
+	if err != nil {
+		return nil, err
+	}
+
+	return &applicationSnapshotEnvironmentBindingList.Items[0], nil
+}
+
+func (a *Adapter) EnsureApplicationSnapshotEnvironmentBindingExist() (results.OperationResult, error) {
+	availableEnvironments, err := a.findAvailabeEnvironments()
+	if err != nil {
+		return results.RequeueWithError(err)
+	}
+
+	applicationSnapshot, err := a.findMatchingApplicationSnapshot()
+	if err != nil {
+		return results.RequeueWithError(err)
+	}
+
+	components, err := a.getAllApplicationComponents(a.application)
+
+	if err != nil {
+		return results.RequeueWithError(err)
+	}
+	for _, availableEnvironment := range *availableEnvironments {
+		bindingName := a.application.Name + "-" + availableEnvironment.Spec.DisplayName + "-" + "binding"
+		applicationSnapshotEnvironmentBinding, _ := a.findExistingApplicationSnapshotEnvironmentBinding(availableEnvironment.Spec.DisplayName)
+		if applicationSnapshotEnvironmentBinding != nil {
+			applicationSnapshotEnvironmentBinding.Spec.Snapshot = applicationSnapshot.Spec.DisplayName
+			applicationSnapshotComponents := a.getApplicationSnapshotComponents(*components)
+			applicationSnapshotEnvironmentBinding.Spec.Components = *applicationSnapshotComponents
+		} else {
+			applicationSnapshotEnvironmentBinding, err := a.CreateApplicationSnapshotEnvironmentBinding(
+				bindingName, a.application.Namespace, a.application.Name,
+				availableEnvironment.Spec.DisplayName,
```

Comment:
We need to set the environment to point to the resource name, so it should be `availableEnvironment.Name` instead of `availableEnvironment.Spec.DisplayName`.

Reply:
Updated

Reply:
Thanks!